### ISSUE TITLE
[Feature Request] Build time hooks?

### ISSUE THREAD

**Issue by Animeshz** [Keats/gutenberg#1931]
I'd like to run a command (like for `postcss`/`parcel-bundler`) every `zola build`.

I know this would work if I just put the css build command before the `zola build`, but when you do `zola serve`, you're also listening to the changes to the directories. In which case multiple builds will be triggerred, and would be good if I'm able to run that command/hook every time that happens.

**Comment by Keats** [comment]
> In which case multiple builds will be triggerred,

Is it that bad? The Zola one should be fast.

**Comment by Animeshz** [comment]
Sorry for ambiguous language... By multiple builds will be triggered I meant one build for each change which is what we'd expect.

I want to hook / know when that happens so post that I could do some cleanup or minification/expansion or whatever etc... :sweat_smile:

**Comment by Keats** [comment]
We should probably find some way to group events rather than reacting to all of them. If we got 10 static assets that changed, we only need to react once, not 10 times.

**Comment by tcaxle** [comment]
I'd also like to see this feature. It wouldn't be too difficult to implement, so I may have a go myself. Here's how I think it should work:

- A `[hooks]` section in the `config.toml` that allows you to define some calls, e.g. `pre-build`, `post-build`, `on-serve` etc.
- Each hook definition is a list
- At appropriate points in the main executable we make OS calls to each of the hooks.

You'd then be able to e.g. make a `hooks/post-build.sh` shell-script that handles your pre-or post processing. For instance you might want to run a script that generates a .thumb.jpg of images you have co-located with your page source files that happens every time you build the site, or to update an ActivityPub outbox/notes file.

Feedback appreciated - I probably don't know what I've let myself in for...

**Comment by burakgungor11235** [comment]
> I'd also like to see this feature. It wouldn't be too difficult to implement, so I may have a go myself. Here's how I think it should work:
> 
>     * A `[hooks]` section in the `config.toml` that allows you to define some calls, e.g. `pre-build`, `post-build`, `on-serve` etc.
> 
>     * Each hook definition is a list
> 
>     * At appropriate points in the main executable we make OS calls to each of the hooks.
> 
> 
> You'd then be able to e.g. make a `hooks/post-build.sh` shell-script that handles your pre-or post processing. For instance you might want to run a script that generates a .thumb.jpg of images you have co-located with your page source files that happens every time you build the site, or to update an ActivityPub outbox/notes file.
> 
> Feedback appreciated - I probably don't know what I've let myself in for...

I agree with you.  A simple `[hooks]` could work. They could define shortcodes like

```toml

[hooks]
max_time = 60 # in seconds
max_mem_per_process = 4 # in Mb
gcc = {binary = "hook_executor.out", associated="gcc"}
```

which in turn will be used like

```md
{{ gcc(src="
// insert generic stuff here
", comp-flags="..."}}
```
which the `hook_executor.out` will be run like this per invocation  

```bash
./hook_executor.out src="..." comp-flags="..."
```

file is executed and the stdout be put into the exact same place as html before the markdown to html process starts by the execution handler. This is sequentional and "dumb". 

```
[file.md] --> [hooks] --> [zola]
```

Would be my guess of how the pipeline runs.  

Once crucial thing is that zola's templating engine does need a `"` substitute. I normally use `@@` and transform it before doing anything else. But that could be delegated to the executor rather than zola. 

This is not fast nor memory efficient but easiest to implement in my eyes. For actual speed we might use a red green algorithm (check hash per invocation) and allowing for dependent and independent hooks so that we have a pool of transformation outputs like 

```md
file1.md:executor[linespan]: executorresult(linespan).html
file1.md:executor[linespan]: executorresult(linespan).html
file1.md:executor[linespan]: executorresult(linespan).html
^^                                                    ^^
Invoked expression at lines   Output of the transformation written in a file 
(will be removed)                     (Will also be removed)   
```

and then a final assembler assembles the file according to the lines in one go rather than multiple transform & assemble steps.  

This harder to implement one would need a pipeline like

```
[files] ---> [Parallel executors] ---> [output pool / markup assembler] ---> [Zola]
```

One note here, if this were to be implemented we'll need to establish a max number of hooks running at the same time so that you don't accidentally spawn a thousand processes each cold build. 

I would also like some feedback on this.
